inc/features/blocks/personalization/edit.js

```diff
@@ -0,0 +1,351 @@
+					}
+				`,
+			} } />
+			<div className={ className }>
+				<div className="wp-core-ui altis-experience-block-header">
+					<span className="altis-experience-block-header__title">
+						{ __( 'Personalized Content', 'altis-experiments' ) }
+						{ ' ・ ' }
+						<VariantTitle variant={ variants[ activeVariantIndex ] } />
+					</span>
+					{ isSelected && (
+						<div className="altis-experience-block-header__toolbar">
+							<IconButton
+								icon='migrate'
+								title={ __( 'Copy variant', 'altis-experiments' ) }
+								disabled={ variants.length < 1 }
+								onClick={ () => {
+									const newClientId = copyVariant( activeVariant );
+									setVariant( newClientId );
+								} }
+							>
+								{ __( 'Copy', 'altis-experiments' ) }
+							</IconButton>
+							{ activeVariant && ! variants[ activeVariantIndex ].attributes.fallback && (
+								<IconButton
+									icon='trash'
+									title={ __( 'Remove variant', 'altis-experiments' ) }
+									disabled={ variants.length < 2 }
+									onClick={ () => {
+										if ( activeVariantIndex === 0 ) {
+											setVariant( variants[ activeVariantIndex + 1 ].clientId );
+										} else {
+											setVariant( variants[ activeVariantIndex - 1 ].clientId );
+										}
+										removeVariant( activeVariant );
+									} }
+								/>
+							) }
+						</div>
+					) }
+				</div>
+				<InnerBlocks
+					allowedBlocks={ ALLOWED_BLOCKS }
+					renderAppender={ false }
+					template={ TEMPLATE }
+				/>
+			</div>
+		</Fragment>
+	);
+};
+
+/**
+ * Creates a new fallback variant block within the provided experience block clientId.
+ *
+ * @param {String} parentId The parent experience block client ID.
+ * @return The new fallback variant block.
+ */
+const createFallbackBlock = parentId => {
+	return createBlock( 'altis/personalization-variant', {
+		parentId,
+		audience: null,
+		fallback: true,
+	} );
+};
+
+export default compose(
+	withSelect( ( select, ownProps ) => {
+		const { clientId, attributes } = ownProps;
+		const { getBlocks } = select( 'core/block-editor' );
+
+		const parentClientId = attributes.clientId || clientId;
+		const innerBlocks = getBlocks( clientId );
+
+		// Ensure at least one variant is present as a fallback.
+		// Note TEMPLATE does not seem to have the desired effect every time.
+		if ( innerBlocks.length === 0 ) {
+			const fallbackBlock = createFallbackBlock( parentClientId );
+			innerBlocks.push( fallbackBlock );
+		} else {
+			// Add a flag to check if we have a fallback explicitly set.
+			let hasFallback = false;
+			innerBlocks.forEach( block => {
+				if ( block.attributes.fallback ) {
+					hasFallback = block.clientId;
+				}
+			} );
+			if ( ! hasFallback ) {
+				const fallbackBlock = createFallbackBlock( parentClientId );
+				innerBlocks.push( fallbackBlock );
+			}
+		}
+
+		return {
+			variants: innerBlocks,
+		};
+	} ),
+	withDispatch( ( dispatch, ownProps, registry ) => {
+		return {
```

Comment:
Could we abstract this code to a separate file to clean it up a bit? It's a lot of complex, untestable code, which isn't tied to the edit component directly